bundles/org.openhab.binding.touchwand/src/main/java/org/openhab/binding/touchwand/internal/TouchWandBridgeHandler.java

```diff
@@ -0,0 +1,139 @@
+import org.openhab.core.thing.binding.ThingHandlerService;
+import org.openhab.core.types.Command;
+import org.osgi.framework.BundleContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link TouchWandBridgeHandler} is responsible for handling commands, which are
+ * sent to one of the channels TouchWand Wanderfull™ Hub channels .
+ *
+ * @author Roie Geron - Initial contribution
+ */
+@NonNullByDefault
+public class TouchWandBridgeHandler extends BaseBridgeHandler implements TouchWandUnitStatusUpdateListener {
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = Collections.singleton(THING_TYPE_BRIDGE);
+    private final Logger logger = LoggerFactory.getLogger(TouchWandBridgeHandler.class);
+    private int statusRefreshRateSec;
+    private boolean addSecondaryUnits;
+    private @Nullable TouchWandWebSockets touchWandWebSockets;
+    private Map<String, TouchWandUnitUpdateListener> unitUpdateListeners = new ConcurrentHashMap<>();
+
+    public TouchWandRestClient touchWandClient;
+
+    public TouchWandBridgeHandler(Bridge bridge, HttpClient httpClient, BundleContext bundleContext) {
+        super(bridge);
+        touchWandClient = new TouchWandRestClient(httpClient);
+        touchWandWebSockets = null;
+    }
+
+    @Override
+    public void initialize() {
+        String host;
+        Integer port;
+        TouchwandBridgeConfiguration config;
+
+        updateStatus(ThingStatus.UNKNOWN);
+
+        config = getConfigAs(TouchwandBridgeConfiguration.class);
+
+        host = config.ipAddress;
+        port = config.port;
+        statusRefreshRateSec = config.statusrefresh;
+        addSecondaryUnits = config.addSecondaryUnits;
+
+        scheduler.execute(() -> {
+            boolean thingReachable = false;
+            String password = config.password;
+            String username = config.username;
+            thingReachable = touchWandClient.connect(username, password, host, port.toString());
+            if (thingReachable) {
+                updateStatus(ThingStatus.ONLINE);
+                TouchWandWebSockets localSockets = touchWandWebSockets = new TouchWandWebSockets(host, scheduler);
+                localSockets.registerListener(this);
+                localSockets.connect();
+            } else {
+                updateStatus(ThingStatus.OFFLINE);
+            }
+        });
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+    }
+
+    public boolean isAddSecondaryControllerUnits() {
+        return addSecondaryUnits;
+    }
+
+    public int getStatusRefreshTime() {
+        return statusRefreshRateSec;
+    }
+
+    @Override
+    public void dispose() {
+        TouchWandWebSockets myTouchWandWebSockets = touchWandWebSockets;
+        if (myTouchWandWebSockets != null) {
+            myTouchWandWebSockets.unregisterListener(this);
+            myTouchWandWebSockets.dispose();
+        }
+    }
```

Comment:
If `dispose` is called immediately after `initialize` your binding should be able to shutdown without side effects.
In this case it is possible for `dispose` to be called before you have initialized the `touchWandWebSockets` field.
To prevent this from happening you need use a boolean `isRunning` field along with synchronization to make sure that access to the `isRunning` and touchWandWebSockets field is atomic.

You could synchronize on a new lock object, but for this fix it is less messy to just synchronize on the `this` object instead.
```
private volatile boolean isRunning = false;

public synchronized void initialize() {
```

```suggestion
        isRunning = true;
        scheduler.execute(() -> {
            boolean thingReachable = false;
            String password = config.password;
            String username = config.username;
            thingReachable = touchWandClient.connect(username, password, host, port.toString());
            if (thingReachable) {
                updateStatus(ThingStatus.ONLINE);
                synchronized(this){
                    if(isRunning){
                        TouchWandWebSockets localSockets = touchWandWebSockets = new TouchWandWebSockets(host, scheduler);
                        localSockets.registerListener(this);
                        localSockets.connect();
                    }
                }
            } else {
                updateStatus(ThingStatus.OFFLINE);
            }
        });
    }

    @Override
    public void handleCommand(ChannelUID channelUID, Command command) {
    }

    public boolean isAddSecondaryControllerUnits() {
        return addSecondaryUnits;
    }

    public int getStatusRefreshTime() {
        return statusRefreshRateSec;
    }

    @Override
    public synchronized void dispose() {
        isRunning = false;
        TouchWandWebSockets myTouchWandWebSockets = touchWandWebSockets;
        if (myTouchWandWebSockets != null) {
            myTouchWandWebSockets.unregisterListener(this);
            myTouchWandWebSockets.dispose();
        }
    }
```

Reply:
done